src/main/java/org/junitpioneer/jupiter/ResourceManagerExtension.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2016-2021 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.jupiter;
+
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+import org.junit.platform.commons.support.ReflectionSupport;
+
+final class ResourceManagerExtension implements ParameterResolver {
+
+	@Override
+	public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext) {
+		// TODO: Also return true if the parameter is annotated with @Shared
+		// TODO: Check that the parameter is not annotated with both @New and @Shared.
+		return parameterContext.isAnnotated(New.class);
+	}
+
+	@Override
+	public Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext)
+			throws ParameterResolutionException {
+		// TODO: Check that the parameter is annotated with at least @New or @Shared.
+		// TODO: Check that the parameter is not annotated with both @New and @Shared.
+		New newResourceAnnotation = parameterContext.findAnnotation(New.class).orElseThrow(() -> {
+			throw new ParameterResolutionException(String
+					.format( //
+						"Parameter `%s` on %s is not annotated with @New", //
+						parameterContext.getParameter(), //
+						extensionContext
+								.getTestMethod()
+								.map(method -> "method `" + method + '`')
+								.orElse("unknown method")));
+		});
+		ResourceFactory<?> resourceFactory =
+				ReflectionSupport.newInstance(newResourceAnnotation.value(), (Object[]) newResourceAnnotation.arguments());
+		extensionContext
+				.getStore(
+					ExtensionContext.Namespace.create(ResourceManagerExtension.class, extensionContext.getUniqueId()))
+				.put(System.identityHashCode(newResourceAnnotation) + "-resource-factory",
```

Comment:
I understand the need for `identityHashCode` as follows:

* the store is a key-value store, so we need a key
* we need to make sure not to accidentally reuse the same key or already-registered resources would get kicked out again

Is that correct? If so, I see one other options besides `identityHashCode`:

* use a random `long` - chances of collisions are minuscule

Reply:
That's correct!

I actually made the same observation a few days ago, as I learned that `System::identityHashCode` [isn't generally a good way of IDing things](https://github.com/plume-lib/plume-util/blob/f7c1a3696ed26dd54a9659bef126a8bf4ea27d9e/src/main/java/org/plumelib/util/UniqueId.java#L9-L14) because when an object is garbage-collected, its memory location can be reused by a new object.

This may not apply in this case, but I think a `long` or `AtomicLong` would be best in case my assumption is ever invalidated.

It's already in the TODO list.